velox/common/memory/SharedArbitrator.cpp

```diff
@@ -169,67 +146,157 @@ const SharedArbitrator::Candidate& findCandidateWithLargestCapacity(
+        freeNonReservedCapacity_,
+        freeReservedCapacity_,
         capacity_,
         toString());
     if (checkUsageLeak_) {
       VELOX_FAIL(errMsg);
     } else {
-      LOG(ERROR) << errMsg;
+      VELOX_MEM_LOG(ERROR) << errMsg;
     }
   }
 }
 
+std::vector<SharedArbitrator::Candidate> SharedArbitrator::getCandidateStats(
+    const std::vector<std::shared_ptr<MemoryPool>>& pools,
+    bool freeCapacityOnly) {
+  std::vector<SharedArbitrator::Candidate> candidates;
+  candidates.reserve(pools.size());
+  for (const auto& pool : pools) {
+    candidates.push_back(
+        {freeCapacityOnly ? 0 : reclaimableUsedCapacity(*pool),
+         reclaimableFreeCapacity(*pool),
+         pool->currentBytes(),
+         pool.get()});
+  }
+  return candidates;
+}
+
+void SharedArbitrator::updateFreeCapacityMetrics() const {
+  RECORD_METRIC_VALUE(
+      kMetricArbitratorFreeCapacityBytes,
+      freeNonReservedCapacity_ + freeReservedCapacity_);
+  RECORD_METRIC_VALUE(
+      kMetricArbitratorFreeReservedCapacityBytes, freeReservedCapacity_);
+}
+
+int64_t SharedArbitrator::reclaimableCapacity(const MemoryPool& pool) const {
+  return std::max<int64_t>(0, pool.capacity() - memoryPoolReservedCapacity_);
+}
+
+int64_t SharedArbitrator::reclaimableFreeCapacity(
+    const MemoryPool& pool) const {
+  return std::min<int64_t>(pool.freeBytes(), reclaimableCapacity(pool));
+}
+
+int64_t SharedArbitrator::reclaimableUsedCapacity(
+    const MemoryPool& pool) const {
+  const auto maxReclaimableBytes = reclaimableCapacity(pool);
+  const auto reclaimableBytes = pool.reclaimableBytes();
+  return std::min<int64_t>(maxReclaimableBytes, reclaimableBytes.value_or(0));
+}
+
+int64_t SharedArbitrator::minGrowCapacity(const MemoryPool& pool) const {
+  return std::max<int64_t>(
+      0,
+      std::min<int64_t>(pool.maxCapacity(), memoryPoolReservedCapacity_) -
+          pool.capacity());
+}
+
 uint64_t SharedArbitrator::growCapacity(
     MemoryPool* pool,
     uint64_t targetBytes) {
-  const int64_t bytesToReserve =
-      std::min<int64_t>(maxGrowBytes(*pool), targetBytes);
-  uint64_t reserveBytes;
-  uint64_t freeCapacity;
+  const auto freeCapacityUpdateCb =
+      folly::makeGuard([this]() { updateFreeCapacityMetrics(); });
+
+  uint64_t reservedBytes{0};
   {
     std::lock_guard<std::mutex> l(mutex_);
     ++numReserves_;
-    reserveBytes = decrementFreeCapacityLocked(bytesToReserve);
-    pool->grow(reserveBytes);
-    freeCapacity = freeCapacity_;
+    const int64_t maxBytesToReserve =
+        std::min<int64_t>(maxGrowCapacity(*pool), targetBytes);
+    const int64_t minBytesToReserve = minGrowCapacity(*pool);
+    reservedBytes =
+        decrementFreeCapacityLocked(maxBytesToReserve, minBytesToReserve);
+    pool->grow(reservedBytes);
+  }
+  return reservedBytes;
+}
+
+uint64_t SharedArbitrator::decrementFreeCapacity(
+    uint64_t maxBytes,
+    uint64_t minBytes) {
```

Comment:
maxReserveBytes, minReserveBytes
or to align with the name maybe maxBytesToReserve, minBytesToReserve